stacker/src/fastcmp.rs

```diff
@@ -0,0 +1,159 @@
+/// fastcmp employs a trick to speed up the comparison of two slices of bytes.
+/// It's also possible to inline compared to the memcmp call.
+///
+/// E.g. Comparing equality of slice length 7 in two steps, by comparing two 4 byte slices
+/// unconditionally instead comparing the remaining 3 bytes if the first comparison was equal.
+/// [1, 2, 3, 4, 5, 6, 7]
+/// [1, 2, 3, 4]
+///          [4, 5, 6, 7]
+///
+/// This method uses the XMM register for bytes slices bigger than 16, else regular registers.
+#[inline]
+pub fn fast_short_slice_compare(left: &[u8], right: &[u8]) -> bool {
+    let len = left.len();
+    if len != right.len() {
+        return false;
+    }
+
+    // This could be less equals, but to make the job a little bit easier for the branch predictor
+    // we put the length 8 into the bigger group (8-16 bytes), that compares two u64
+    // assuming that range 8-16 are more common than 4-7
+
+    // This weird branching is done on purpose to get the best assembly.
+    // if len< 4 {
+    // ..
+    // if len < 8
+    // will cause assembly inlined instead of jumps
+    if len < 8 {
+        if len >= 4 {
+            return double_check_trick::<4>(left, right);
+        } else {
+            return short_compare(left, right);
+        }
+    }
+
+    if len > 16 {
+        return fast_nbyte_slice_compare::<16>(left, right);
+    }
+
+    double_check_trick::<8>(left, right)
+}
+
+#[inline]
+// Note: The straigthforward left.chunks_exact(SIZE).zip(right.chunks_exact(SIZE)) produces slower
+// assembly
+fn fast_nbyte_slice_compare<const SIZE: usize>(mut left: &[u8], mut right: &[u8]) -> bool {
+    let left_orig = left;
+    let right_orig = right;
+
+    loop {
+        if left.len() < SIZE || right.len() < SIZE {
+            break;
+        }
```

Comment:
```suggestion
        assert_eq!(left.len(), right.len()); //< I assume this will go away after inlining?
        while left.len() >= SIZE { //< I assume this one cannot go away after inlining? (but maybe LLVM is stronger than I think)
```
    while left.len() >= SIZE {
```

Reply:
I confirmed with godbolt this does two checks with the current code.

Reply:
This does not remove the bounds checks.
It seems that the compiler can't deduce from same length + same increments on left and right to remove bounds checks

Reply:
hmm I checked with godbolt and that's not true...